[错误报告]: pip依赖安装后的模块刷新误删运行态模块，导致微信/CookieCloud AES解密失败

### 确认

- [x] 我的版本是最新版本，我的版本号与 [version](https://github.com/jxxghp/MoviePilot/releases/latest) 相同。
- [x] 我已经 [issue](https://github.com/jxxghp/MoviePilot/issues) 中搜索过，确认我的问题没有被提出过。
- [x] 我已经 [Telegram频道](https://t.me/moviepilot_channel) 中搜索过，确认我的问题没有被提出过。
- [x] 我已经修改标题，将标题中的 描述 替换为我遇到的问题。

### 当前程序版本

2.10.8

### 运行环境

Docker

### 问题类型

主程序运行问题

### 问题描述

## 问题现象

在 Docker 环境启动后，出现以下稳定异常：
1. 企业微信消息解密失败：`DecryptMsg ret = -40007`
2. CookieCloud 同步失败：`multiple declarations of function CBC_start_operation`

两者都发生在启动期插件依赖安装完成后，且可在同一进程中连续复现。

## 关联issue
#4610
#4538

## 复现条件

- Docker 环境，已安装"动态企微可信IP"插件
- 启动过程中触发插件依赖安装（`/config/temp/plugin_dependencies/requirements.txt`）
- 同时存在插件导入/热加载行为（日志中可见 `app.plugins.dynamicwechat` 导入）
- 随后测试：
  - 企业微信交互（触发解密）
  - CookieCloud 同步（触发 AES 解密）

## 关键日志

- pip 安装后刷新导入系统时，`new_modules` 包含：
  - `app.plugins.dynamicwechat`
  - `app.plugins.dynamicwechat.helper`
  - `Crypto.Cipher._mode_cbc`
- 后续出现：
  - `DecryptMsg ret = -40007`
  - `multiple declarations of function CBC_start_operation`

## 原因分析

当前 `pip_install_with_fallback()` 在安装成功后采用如下策略：

- 记录 `before_installation = set(sys.modules)`
- 安装后计算 `new_modules = current_modules - before_installation`
- 对 `new_modules` 执行 `del sys.modules[module]`

该逻辑假设 `new_modules` 仅包含“pip新增模块”。  
但在多线程/并发导入场景下，`new_modules` 会混入同时间窗内其他线程导入的业务模块及其依赖（本次为 `dynamicwechat -> Crypto.Cipher._mode_cbc`）。  
最终导致运行态模块被误删，后续 AES/CBC 路径再次加载时触发底层重复声明异常。

## 修复建议

1. **移除 `del sys.modules[...]` 差集删除机制**
   - 保留 `importlib.invalidate_caches()`（可选保留 `importlib.reload(site)`）
2. **为 pip 安装流程加全局锁**
   - 串行化 `pip_install_with_fallback`，避免安装流程并发互踩
3. **安装期间抑制插件热重载/并发导入**
   - 降低导入窗口污染概率
4. **优化日志**
   - 不再打印超大 `new_modules` 集合，改为摘要输出

## 预期结果

- 启动期依赖安装后，不再出现 `Crypto.Cipher._mode_cbc` 相关误删
- 微信解密与 CookieCloud 解密恢复正常
- 插件依赖安装流程在并发场景下稳定可预测

### 发生问题时系统日志和配置文件

```bash
2026-04-30 10:59:52 [DEBUG] plugin.py - [PIP] 策略：镜像站 安装依赖成功
  requirements: /config/temp/plugin_dependencies/requirements.txt
  installed: cacheout, python_hosts, webdavclient3, apprise

2026-04-30 10:59:52 [DEBUG] plugin.py - [PIP] 已刷新导入系统，新加载的模块: {..., 'app.plugins.dynamicwechat', 'app.plugins.dynamicwechat.helper', 'Crypto.Cipher._mode_cbc', ...}
2026-04-30 10:59:52 [INFO ] plugin.py - 已完成 4 个依赖项安装，总耗时：10.06 秒

2026-04-30 11:00:29 [ERROR] plugin.py - [PIP] 策略：镜像站 安装依赖失败
  TypeError: expected string or bytes-like object, got 'NoneType'
  BadMetadata: invalid metadata entry `version`
2026-04-30 11:00:29 [DEBUG] plugin.py - [PIP] 尝试使用策略：直连 安装依赖
2026-04-30 11:01:01 [DEBUG] plugin.py - [PIP] 策略：直连 安装依赖成功

2026-04-30 11:02:00 [DEBUG] wechat - 收到微信请求: [REDACTED_WECHAT_XML]
2026-04-30 11:02:00 [ERROR] wechat - 解密微信消息失败 DecryptMsg ret = -40007

2026-04-30 11:02:30 [INFO ] site.py - 开始同步CookieCloud站点 ...
2026-04-30 11:02:30 [ERROR] site.py - CookieCloud同步失败：cookie解密失败：multiple declarations of function CBC_start_operation (for interactive usage, try cdef(xx, override=True))
```

--- a/app/helper/plugin.py
+++ b/app/helper/plugin.py
@@ -5,6 +5,7 @@ import json
 import shutil
 import site
 import sys
+import threading
 import traceback
 import zipfile
 from pathlib import Path
@@ -45,6 +46,8 @@ class PluginHelper(metaclass=WeakSingleton):
     _install_reg = f"{settings.MP_SERVER_HOST}/plugin/install/{{pid}}"
     _install_report = f"{settings.MP_SERVER_HOST}/plugin/install"
     _install_statistic = f"{settings.MP_SERVER_HOST}/plugin/statistic"
+    # 串行化运行期依赖安装，避免多个 pip 子进程和导入缓存刷新互相踩踏。
+    _pip_install_lock = threading.Lock()
 
     def __init__(self):
         self.systemconfig = SystemConfigOper()
@@ -873,28 +876,24 @@ class PluginHelper(metaclass=WeakSingleton):
             strategies.append(("代理", base_cmd + ["--proxy", settings.PROXY_HOST]))
         strategies.append(("直连", base_cmd))
 
-        # 记录当前已安装的包，以便后续刷新
-        before_installation = set(sys.modules.keys())
-
-        # 遍历策略进行安装
-        for strategy_name, pip_command in strategies:
-            logger.debug(f"[PIP] 尝试使用策略：{strategy_name} 安装依赖，命令：{' '.join(pip_command)}")
-            success, message = SystemUtils.execute_with_subprocess(pip_command)
-            if success:
-                logger.debug(f"[PIP] 策略：{strategy_name} 安装依赖成功，输出：{message}")
-                # 安装成功后刷新Python的模块系统
-                importlib.reload(site)
-                # 获取新安装的模块
-                current_modules = set(sys.modules.keys())
-                new_modules = current_modules - before_installation
-                # 重新加载新安装的模块
-                for module in new_modules:
-                    if module in sys.modules:
-                        del sys.modules[module]
-                logger.debug(f"[PIP] 已刷新导入系统，新加载的模块: {new_modules}")
-                return True, message
-            else:
-                logger.error(f"[PIP] 策略：{strategy_name} 安装依赖失败，错误信息：{message}")
+        # pip 会修改当前解释器的 site-packages，安装与缓存刷新必须串行，避免运行态模块被并发安装窗口污染。
+        with PluginHelper._pip_install_lock:
+            loaded_modules_before_install = set(sys.modules.keys())
+            # 遍历策略进行安装
+            for strategy_name, pip_command in strategies:
+                logger.debug(f"[PIP] 尝试使用策略：{strategy_name} 安装依赖，命令：{' '.join(pip_command)}")
+                success, message = SystemUtils.execute_with_subprocess(pip_command)
+                if success:
+                    logger.debug(f"[PIP] 策略：{strategy_name} 安装依赖成功，输出：{message}")
+                    # 刷新导入系统即可发现新安装依赖，同时保持安装窗口内的运行态模块缓存稳定。
+                    importlib.reload(site)
+                    importlib.invalidate_caches()
+                    loaded_modules_after_install = set(sys.modules.keys())
+                    loaded_modules_during_install = loaded_modules_after_install - loaded_modules_before_install
+                    logger.debug(f"[PIP] 已刷新导入系统，新加载的模块: {loaded_modules_during_install}")
+                    return True, message
+                else:
+                    logger.error(f"[PIP] 策略：{strategy_name} 安装依赖失败，错误信息：{message}")
 
         return False, "[PIP] 所有策略均安装依赖失败，请检查网络连接或 PIP 配置"
 

--- a/tests/test_plugin_helper.py
+++ b/tests/test_plugin_helper.py
@@ -1,4 +1,11 @@
+import sys
+import tempfile
+import threading
+import time
+from pathlib import Path
+from types import ModuleType
 from unittest import TestCase
+from unittest.mock import patch
 
 
 class PluginHelperTest(TestCase):
@@ -21,3 +28,92 @@ class PluginHelperTest(TestCase):
             "local://TestPlugin?version=v2",
             PluginHelper.sanitize_repo_url_for_statistic(repo_url)
         )
+
+    def test_pip_install_keeps_modules_imported_during_install(self):
+        """
+        验证依赖安装窗口内被其他任务导入的运行态模块不会被误删。
+        """
+        try:
+            from app.helper.plugin import PluginHelper
+        except ModuleNotFoundError as exc:
+            self.skipTest(f"missing dependency: {exc}")
+
+        module_names = ["app.plugins.dynamicwechat.helper", "Crypto.Cipher._mode_cbc"]
+        previous_modules = {name: sys.modules.get(name) for name in module_names}
+
+        def fake_execute(_cmd):
+            for module_name in module_names:
+                sys.modules[module_name] = ModuleType(module_name)
+            return True, "ok"
+
+        try:
+            with tempfile.TemporaryDirectory() as temp_dir:
+                requirements_file = Path(temp_dir) / "requirements.txt"
+                requirements_file.write_text("demo-package\n", encoding="utf-8")
+                with patch("app.helper.plugin.SystemUtils.execute_with_subprocess", side_effect=fake_execute):
+                    success, message = PluginHelper.pip_install_with_fallback(requirements_file)
+
+            self.assertTrue(success)
+            self.assertEqual("ok", message)
+            for module_name in module_names:
+                self.assertIn(module_name, sys.modules)
+        finally:
+            for module_name, previous_module in previous_modules.items():
+                if previous_module is None:
+                    sys.modules.pop(module_name, None)
+                else:
+                    sys.modules[module_name] = previous_module
+
+    def test_pip_install_serializes_concurrent_calls(self):
+        """
+        验证多个依赖安装请求会复用同一把锁串行执行 pip。
+        """
+        try:
+            from app.helper.plugin import PluginHelper
+        except ModuleNotFoundError as exc:
+            self.skipTest(f"missing dependency: {exc}")
+
+        thread_count = 2
+        active_installs = 0
+        max_active_installs = 0
+        state_lock = threading.Lock()
+        start_event = threading.Event()
+        errors = []
+
+        def fake_execute(_cmd):
+            nonlocal active_installs, max_active_installs
+            with state_lock:
+                active_installs += 1
+                max_active_installs = max(max_active_installs, active_installs)
+            time.sleep(0.05)
+            with state_lock:
+                active_installs -= 1
+            return True, "ok"
+
+        def worker(requirements_file: Path):
+            try:
+                start_event.wait()
+                PluginHelper.pip_install_with_fallback(requirements_file)
+            except Exception as err:  # pragma: no cover - 仅用于并发测试失败诊断
+                errors.append(err)
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            requirements_files = []
+            for index in range(thread_count):
+                requirements_file = Path(temp_dir) / f"requirements-{index}.txt"
+                requirements_file.write_text("demo-package\n", encoding="utf-8")
+                requirements_files.append(requirements_file)
+
+            threads = [
+                threading.Thread(target=worker, args=(requirements_file,))
+                for requirements_file in requirements_files
+            ]
+            with patch("app.helper.plugin.SystemUtils.execute_with_subprocess", side_effect=fake_execute):
+                for thread in threads:
+                    thread.start()
+                start_event.set()
+                for thread in threads:
+                    thread.join()
+
+        self.assertEqual([], errors)
+        self.assertEqual(1, max_active_installs)

